### ISSUE TITLE
Azure AKS Files-backed PVC: unbound immediate PVC errors

### ISSUE THREAD

**Issue by ryedin** [argoproj/argo#4003]
## Summary 

Workflow pods fail to bind to the PVC properly when using a storage class that is backed by Azure Files. I do not know if that detail is specifically relevant, but it certainly appears that way. The storage account and file share and volume and the PVC are all created correctly, but the pods simply fail to actually bind, and we get `pod has unbound immediate PersistentVolumeClaims (repeated 2 times)`

Using the exact same PVC spec against a non-argo deployment pod, there are no issues.

## Diagnostics

What version of Argo Workflows are you running?

`v2.10.1`

```yaml
kind: Workflow
apiVersion: argoproj.io/v1alpha1
metadata:
  creationTimestamp: '2020-09-11T16:34:19Z'
  generation: 5
  labels:
    workflows.argoproj.io/archive-strategy: 'false'
    workflows.argoproj.io/completed: 'true'
    workflows.argoproj.io/phase: Error
  name: etl-workflow-test2-5
  namespace: hoonuit-etl
  resourceVersion: '208868'
  selfLink: >-
    /apis/argoproj.io/v1alpha1/namespaces/hoonuit-etl/workflows/etl-workflow-test2-5
  uid: 27788787-1489-4fb3-89ba-77fc2e6a78ef
spec:
  arguments: {}
  entrypoint: run
  imagePullSecrets:
    - name: registry-pull-secret
  templates:
    - arguments: {}
      dag:
        tasks:
          - arguments: {}
            name: execute
            template: execute
      inputs: {}
      metadata: {}
      name: run
      outputs: {}
    - arguments: {}
      container:
        args:
          - foo
        command:
          - cowsay
        image: docker/whalesay
        name: ''
        resources: {}
        volumeMounts:
          - mountPath: /usr/data
            name: etl-data
      inputs: {}
      metadata:
        annotations:
          sidecar.istio.io/inject: 'false'
      name: execute
      outputs: {}
  volumeClaimTemplates:
    - metadata:
        creationTimestamp: null
        name: etl-data
      spec:
        accessModes:
          - ReadWriteMany
        resources:
          requests:
            storage: 6Gi
        storageClassName: etl-azurefile-storage-class
      status: {}
status:
  conditions:
    - status: 'True'
      type: Completed
  finishedAt: '2020-09-11T16:34:38Z'
  nodes:
    etl-workflow-test2-5:
      children:
        - etl-workflow-test2-5-1522985522
      displayName: etl-workflow-test2-5
      finishedAt: '2020-09-11T16:34:38Z'
      id: etl-workflow-test2-5
      name: etl-workflow-test2-5
      outboundNodes:
        - etl-workflow-test2-5-1522985522
      phase: Error
      startedAt: '2020-09-11T16:34:19Z'
      templateName: run
      templateScope: local/etl-workflow-test2-5
      type: DAG
    etl-workflow-test2-5-1522985522:
      boundaryID: etl-workflow-test2-5
      displayName: execute
      finishedAt: '2020-09-11T16:34:36Z'
      hostNodeName: aks-etl01-14543412-vmss000001
      id: etl-workflow-test2-5-1522985522
      message: >-
        failed to save outputs: Failed to establish pod watch: timed out waiting
        for the condition
      name: etl-workflow-test2-5.execute
      phase: Error
      resourcesDuration:
        cpu: 7
        memory: 7
      startedAt: '2020-09-11T16:34:19Z'
      templateName: execute
      templateScope: local/etl-workflow-test2-5
      type: Pod
  phase: Error
  resourcesDuration:
    cpu: 7
    memory: 7
  startedAt: '2020-09-11T16:34:19Z'


```

relevent log lines:

```
2020-09-11T16:34:19.021187017Z time="2020-09-11T16:34:19Z" level=info msg="Processing workflow" namespace=hoonuit-etl workflow=etl-workflow-test2-5
2020-09-11T16:34:19.021234618Z time="2020-09-11T16:34:19Z" level=info msg="Updated phase  -> Running" namespace=hoonuit-etl workflow=etl-workflow-test2-5
2020-09-11T16:34:19.021832121Z time="2020-09-11T16:34:19Z" level=info msg="Creating pvc etl-workflow-test2-5-etl-data" namespace=hoonuit-etl workflow=etl-workflow-test2-5
2020-09-11T16:34:19.034549095Z time="2020-09-11T16:34:19Z" level=info msg="DAG node etl-workflow-test2-5 initialized Running" namespace=hoonuit-etl workflow=etl-workflow-test2-5
2020-09-11T16:34:19.034593595Z time="2020-09-11T16:34:19Z" level=info msg="All of node etl-workflow-test2-5.execute dependencies [] completed" namespace=hoonuit-etl workflow=etl-workflow-test2-5
2020-09-11T16:34:19.034770196Z time="2020-09-11T16:34:19Z" level=info msg="Pod node etl-workflow-test2-5-1522985522 initialized Pending" namespace=hoonuit-etl workflow=etl-workflow-test2-5
2020-09-11T16:34:19.174628305Z time="2020-09-11T16:34:19Z" level=info msg="Created pod: etl-workflow-test2-5.execute (etl-workflow-test2-5-1522985522)" namespace=hoonuit-etl workflow=etl-workflow-test2-5
2020-09-11T16:34:19.182420250Z time="2020-09-11T16:34:19Z" level=info msg="insignificant pod change" key=hoonuit-etl/etl-workflow-test2-5-1522985522
2020-09-11T16:34:19.183775858Z time="2020-09-11T16:34:19Z" level=info msg="Workflow update successful" namespace=hoonuit-etl phase=Running resourceVersion=208802 workflow=etl-workflow-test2-5
2020-09-11T16:34:20.184471849Z time="2020-09-11T16:34:20Z" level=info msg="Processing workflow" namespace=hoonuit-etl workflow=etl-workflow-test2-5
2020-09-11T16:34:20.192705997Z time="2020-09-11T16:34:20Z" level=info msg="Updating node etl-workflow-test2-5-1522985522 message: Unschedulable: pod has unbound immediate PersistentVolumeClaims (repeated 2 times)"
2020-09-11T16:34:20.234466438Z time="2020-09-11T16:34:20Z" level=info msg="Skipped pod etl-workflow-test2-5.execute (etl-workflow-test2-5-1522985522) creation: already exists" namespace=hoonuit-etl workflow=etl-workflow-test2-5
2020-09-11T16:34:20.242206483Z time="2020-09-11T16:34:20Z" level=info msg="Workflow update successful" namespace=hoonuit-etl phase=Running resourceVersion=208812 workflow=etl-workflow-test2-5
2020-09-11T16:34:21.242907174Z time="2020-09-11T16:34:21Z" level=info msg="Processing workflow" namespace=hoonuit-etl workflow=etl-workflow-test2-5
2020-09-11T16:34:21.247866203Z time="2020-09-11T16:34:21Z" level=info msg="Updating node etl-workflow-test2-5-1522985522 message: ContainerCreating"
2020-09-11T16:34:21.303986628Z time="2020-09-11T16:34:21Z" level=info msg="Skipped pod etl-workflow-test2-5.execute (etl-workflow-test2-5-1522985522) creation: already exists" namespace=hoonuit-etl workflow=etl-workflow-test2-5
2020-09-11T16:34:21.311639372Z time="2020-09-11T16:34:21Z" level=info msg="Workflow update successful" namespace=hoonuit-etl phase=Running resourceVersion=208818 workflow=etl-workflow-test2-5
2020-09-11T16:34:22.312166862Z time="2020-09-11T16:34:22Z" level=info msg="Processing workflow" namespace=hoonuit-etl workflow=etl-workflow-test2-5
2020-09-11T16:34:22.377377039Z time="2020-09-11T16:34:22Z" level=info msg="Skipped pod etl-workflow-test2-5.execute (etl-workflow-test2-5-1522985522) creation: already exists" namespace=hoonuit-etl workflow=etl-workflow-test2-5
2020-09-11T16:34:38.064247215Z time="2020-09-11T16:34:38Z" level=info msg="Processing workflow" namespace=hoonuit-etl workflow=etl-workflow-test2-5
2020-09-11T16:34:38.069375345Z time="2020-09-11T16:34:38Z" level=info msg="Updating node etl-workflow-test2-5-1522985522 status Pending -> Error"
2020-09-11T16:34:38.069399145Z time="2020-09-11T16:34:38Z" level=info msg="Updating node etl-workflow-test2-5-1522985522 message: failed to save outputs: Failed to establish pod watch: timed out waiting for the condition"
2020-09-11T16:34:38.069668846Z time="2020-09-11T16:34:38Z" level=info msg="Outbound nodes of etl-workflow-test2-5 set to [etl-workflow-test2-5-1522985522]" namespace=hoonuit-etl workflow=etl-workflow-test2-5
2020-09-11T16:34:38.069690046Z time="2020-09-11T16:34:38Z" level=info msg="node etl-workflow-test2-5 phase Running -> Error" namespace=hoonuit-etl workflow=etl-workflow-test2-5
2020-09-11T16:34:38.069695846Z time="2020-09-11T16:34:38Z" level=info msg="node etl-workflow-test2-5 finished: 2020-09-11 16:34:38.069529245 +0000 UTC" namespace=hoonuit-etl workflow=etl-workflow-test2-5
2020-09-11T16:34:38.069700246Z time="2020-09-11T16:34:38Z" level=info msg="Checking daemoned children of etl-workflow-test2-5" namespace=hoonuit-etl workflow=etl-workflow-test2-5
2020-09-11T16:34:38.069704846Z time="2020-09-11T16:34:38Z" level=info msg="Deleting PVC etl-workflow-test2-5-etl-data" namespace=hoonuit-etl workflow=etl-workflow-test2-5
2020-09-11T16:34:38.076570586Z time="2020-09-11T16:34:38Z" level=info msg="Deleted 1/1 PVCs" namespace=hoonuit-etl workflow=etl-workflow-test2-5
2020-09-11T16:34:38.076613886Z time="2020-09-11T16:34:38Z" level=info msg="Updated phase Running -> Error" namespace=hoonuit-etl workflow=etl-workflow-test2-5
2020-09-11T16:34:38.076627487Z time="2020-09-11T16:34:38Z" level=info msg="Marking workflow completed" namespace=hoonuit-etl workflow=etl-workflow-test2-5
2020-09-11T16:34:38.076637587Z time="2020-09-11T16:34:38Z" level=info msg="Checking daemoned children of " namespace=hoonuit-etl workflow=etl-workflow-test2-5
2020-09-11T16:34:38.084534432Z time="2020-09-11T16:34:38Z" level=info msg="Workflow update successful" namespace=hoonuit-etl phase=Error resourceVersion=208868 workflow=etl-workflow-test2-5
2020-09-11T16:34:39.084876721Z time="2020-09-11T16:34:39Z" level=info msg="etl-workflow-test2-5 released all acquired locks"
2020-09-11T16:34:39.093438570Z time="2020-09-11T16:34:39Z" level=info msg="Workflow update successful" namespace=hoonuit-etl phase=Error resourceVersion=208868 workflow=etl-workflow-test2-5
2020-09-11T16:34:39.093931473Z time="2020-09-11T16:34:39Z" level=info msg="Labeled pod hoonuit-etl/etl-workflow-test2-5-1522985522 completed"
```

StorageClass definition:

```yaml
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: etl-azurefile-storage-class
provisioner: kubernetes.io/azure-file
parameters:
  skuName: Standard_LRS
  secretNamespace: hoonuit-etl
```

Again, it's worth re-iterating that when I manually create a PVC and mount that to a regular deployment, it just works. So it looks to me like it's clearly an argo bug, or I'm missing some config param to tell argo to wait longer for the PVCs to bind? Obviously using a normal task-level timeout is not right either because that has to do with the time it takes for the actual work to be done.

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by ryedin** [comment]
As a quick update to this. Some time (maybe a minute or two) after argo fails the workflow, the PVC does get bound to the (now completed) pod.

So, how can I tell the workflow controller to chill for a bit and actually wait for PVCs to bind before failing the workflow?

**Comment by sarabala1979** [comment]
You can add `entry step` to check the PVC status or have a `suspend template` to sleep the workflow for a given time period.

**Comment by sarabala1979** [comment]
it looks like the below error is caused to fail the workflow, not for PVC pending.
```
2020-09-11T16:34:38.069375345Z time="2020-09-11T16:34:38Z" level=info msg="Updating node etl-workflow-test2-5-1522985522 status Pending -> Error"
2020-09-11T16:34:38.069399145Z time="2020-09-11T16:34:38Z" level=info msg="Updating node etl-workflow-test2-5-1522985522 message: failed to save outputs: Failed to establish pod watch: timed out waiting for the condition
```

Can you provide the pod's logs for the `main` and `wait` container?
`kubectl logs <pods> main`
`kubectl logs <pods> wait`

**Comment by ryedin** [comment]
... ok so yeah this turns out to have been a big red herring.

digging into the `wait` container logs (which didn't even occur to me at first tbh) revealed that the real problem was RBAC. (The default service account could not list the pods). My local testing up to this point was in a non-RBAC cluster and I missed the obvious due to what I perceived as being the only variable (file shares vs hostpath for the PVCs), combined with the red herring errors I saw and the fact that I was getting no actual feedback about this condition in the UI...

I added the appropriate RBAC resources and the workflow now runs. However I'm now having a different issue with the UI not actually rendering anything for the workflow. I'll dig through the issues and try to isolate that one and make a suitable ticket for that if I think it's warranted.

Thank you, sorry for the noise. I'll close this one.
